lib/srv/desktop/ldap.go

```diff
@@ -17,64 +17,43 @@ limitations under the License.
 package desktop
 
 import (
-	"crypto/tls"
-	"crypto/x509"
 	"fmt"
 	"strings"
+	"sync"
 
 	"github.com/go-ldap/ldap/v3"
 	"github.com/gravitational/trace"
 )
 
 // Note: if you want to browse LDAP on the Windows machine, run ADSIEdit.msc.
 type ldapClient struct {
-	cfg    LDAPConfig
+	cfg LDAPConfig
+
+	mu     sync.Mutex
 	client ldap.Client
 }
 
-// newLDAPClient connects to an LDAP server, authenticates and returns the
-// client connection. Caller must close the client after using it.
-func newLDAPClient(cfg LDAPConfig) (*ldapClient, error) {
-	tlsConfig := &tls.Config{
-		InsecureSkipVerify: cfg.InsecureSkipVerify,
-	}
-
-	if !cfg.InsecureSkipVerify {
-		// Get the SystemCertPool, continue with an empty pool on error
-		rootCAs, _ := x509.SystemCertPool()
-		if rootCAs == nil {
-			rootCAs = x509.NewCertPool()
-		}
-
-		if cfg.CA != nil {
-			// Append our cert to the pool.
-			rootCAs.AddCert(cfg.CA)
-		}
-
-		// Supply our cert pool to TLS config for verification.
-		tlsConfig.RootCAs = rootCAs
-	}
-
-	con, err := ldap.DialURL("ldaps://"+cfg.Addr, ldap.DialWithTLSConfig(tlsConfig))
-	if err != nil {
-		return nil, trace.Wrap(err)
+func (c *ldapClient) setClient(client ldap.Client) {
+	c.mu.Lock()
+	if c.client != nil {
+		c.client.Close()
```

Comment:
If a client is returned by `ldapClient.c()` and then the setClient will be  called the next use of the client returned by the `ldapClient.c()` will  failed because the client was closed.